bam_samples.c

```diff
@@ -0,0 +1,369 @@
+static int load_dictionaries(Params* params, const char* filename) {
+    int ret;
+    htsFile* in;
+    int status = EXIT_SUCCESS;
+    kstring_t  *ks = NULL;
+    in = hts_open(filename,"r");
+
+    ks = &in->line;
+    if (in == NULL) {
+        print_error_errno("samples","cannot open \"%s\".", filename);
+        return EXIT_FAILURE;
+    }
+    while ((ret = hts_getline(in, KS_SEP_LINE, ks)) >= 0) {
+        if (load_dictionary(params, ks->s)!= EXIT_SUCCESS) {
+            status = EXIT_FAILURE;
+            break;
+        }
+    }
+    hts_close(in);
+    return status;
+}
+
+/** print the sample information, search for a reference */
+static int print_sample(Params* params, sam_hdr_t *header, int has_index, const char* sample, const char* fname) {
+    fputs(sample, params->out);
+    fputc('\t', params->out);
+    fputs(fname, params->out);
+    if (params->test_index) {
+        fprintf(params->out, "\t%d", has_index);
+    }
+    if (params->faidx!=NULL) {
+        FaidxPath* ref = NULL;
+        FaidxPath* curr = params->faidx;
+        while(curr!=NULL) {
+            /** check names and length are the same in the same order */
+            if (faidx_nseq(curr->faidx) == header->n_targets) {
+                int i;
+                for (i = 0; i < faidx_nseq(curr->faidx); i++) {
+                    /** check name if the same */
+                    if (strcmp(faidx_iseq(curr->faidx, i), header->target_name[i])!=0) break;
+                    /** check length if the same */
+                    if (faidx_seq_len(curr->faidx, faidx_iseq(curr->faidx, i)) != header->target_len[i]) break;
+                }
+                if (i == faidx_nseq(curr->faidx)) {
+                    ref = curr;
+                    break;
+                }
+            }
+            curr = curr->next;
+        }
+        fputc('\t', params->out);
+        if (ref == NULL) {
+            fputc('.', params->out);
+        } else {
+            fputs(curr->filename, params->out);
+        }
+    }
+    fputc('\n', params->out);
+    return 0;
+}
+
+
+/** open a sam file. Search for all samples in the @RG lines */
+static int print_samples(Params* params, const char* fname) {
+    samFile *in = 0;
+    sam_hdr_t *header = NULL;
+    int n_rg;
+    int status = EXIT_SUCCESS;
+    khash_t(sm) *sample_set = kh_init(sm);
+    khint_t k;
+    int count_samples = 0;
+    int has_index = 0;
+
+
+    /* open sam file */
+    if ((in = sam_open_format(fname, "r", NULL)) == 0) {
+        print_error_errno("samples", "failed to open \"%s\" for reading.", fname);
+        status = EXIT_FAILURE;
+        goto end_print;
+    }
+    /* load header */
+    if ((header = sam_hdr_read(in)) == 0) {
+        print_error("samples", "failed to read the header from \"%s\".", fname);
+        status = EXIT_FAILURE;
+        goto end_print;
+    }
+
+    /* try to load index if required */
+    if (params->test_index) {
+        hts_idx_t *bam_idx =  sam_index_load(in, fname);
```

Comment:
The code should accommodate index file names that are not prefixed by the alignment file name. There could be `my_index.bam.bai` as an index for `alignments.bam`, instead of `alignments.bam.bai`. Perhaps the `-i` option should take an argument.